### DISCUSSION TITLE
Руководство ПЛИ, п.2.1. "Общие ограничения"

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#267] [Лабораторная РГ]
Рассматриваем перевод на FHIR пункта 2.1. Руководства ПЛИ, редакция 4.
<img width="493" alt="image" src="https://github.com/user-attachments/assets/8d88a977-a5ef-4a3a-926b-377cdb157156">

**Comment by eugenekogan-spb** [comment]
2.1.1. Имена и наименования
<img width="651" alt="image" src="https://github.com/user-attachments/assets/9c54ffb6-02c7-4162-b643-c7737d90afd3">

Пример указания имени автора и названия организации
<img width="763" alt="image" src="https://github.com/user-attachments/assets/07f82292-0ef9-4846-bd4c-0dfd113b4a73">

**Reply by eugenekogan-spb** [reply]
Анализ
FHIR предлагает тип данных HumanName
<img width="830" alt="image" src="https://github.com/user-attachments/assets/522df72b-c970-4b0e-adac-3747c3c99933">

Для реализации требований СЭМД ПЛИ можно применить к этому типу данных следующие ограничения:
family 1..1 //фамилия обязательна
given 1..2 //имя обязательно, отчество может быть указано

Для приведенного примера (если бы это был пациент)
```
Instance:   pli-patient-instance-name
InstanceOf: Pli_Patient
Title: "Pli Instance Patient with name"
Description: "Пациент с именем"
Usage: #example

* name
  * family = "Новосельцев"
  * given[0] = "Михаил"
  * given[1] = "Владимирович"

```

В ресурсе Organization есть необязательное поле для названия
<img width="567" alt="image" src="https://github.com/user-attachments/assets/390bebde-bfd5-4a00-990e-25522da3b165">
для СЭМД ПЛИ требуется указать обязательность:

name 1..1

Пример
```
Instance: core-organization-instance-full
InstanceOf: Core_Organization
Title: "Core Instance Organization Full"
Description: "Психиатрическая больница им.Кащенко с адресом и идентификаторами ФРМР, ИНН, ОГРН"
Usage: #example

* name = "Государственное бюджетное учреждение города Москвы "Городская поликлиника №129 Департамента здравоохранения города Москвы" "
```

**Reply by eugenekogan-spb** [reply]
Для атрибута "Отчество":
* В СЭМД - расширение.
* В FHIR - второй элемент массива.

Включаем в реестр расхождений.
Информация в FHIR-профиле достаточна, дополнений не требуется.

**Comment by eugenekogan-spb** [comment]
2.1.2 Адреса
<img width="633" alt="image" src="https://github.com/user-attachments/assets/307078a9-d6a7-4326-b245-817ba30c905f">

**Reply by eugenekogan-spb** [reply]
Анализ

FHIR предлагает тип данных Address
<img width="876" alt="image" src="https://github.com/user-attachments/assets/08dd653f-bc4e-4d16-aab5-ecf4d95940f4">
Требования СЭМД ПЛИ можно выразить следующим образом:
1. Строковое представление обязательно
text 1..1 

2. Обязан иметь элемент stateCode
Можно определить Extension 
regionRF 1..1 coding, заполняется по справочнику субъектов РФ, 1.2.643.5.1.13.13.99.2.206

3. должен иметь PostalCode
Атрибут postalCode есть, при отсутствии значения необходимо указывать nullFlavor

4. должен иметь элемент fias:Address
Есть определенное в RuCore расширение fias, необходимо указать, что при отсутствии значения необходимо заполнить nullFlavor

Пример адреса, соответствующий приведенному в спецификации СЭМД ПЛИ:
```
Instance:   pli-address-instance
InstanceOf: Pli_Address
Title: "Core Instance Pli"
Description: "Заполнение адреса как в СЭМД"
Usage: #inline

* use = #home

* type = #postal

* text = "Московская область, город Москва, улица Ладожская, дом 4/6"

* postalCode = "105005"

* extension[regionRF]
  * valueCoding
      * code = "77"
      * system = "1.2.643.5.1.13.13.99.2.20"
      * version = "6.3"         
      * display = "г. Москва"

* extension[fias]
  * extension[aoguid]
    * valueIdentifier
      * value = "212d3123-9c4d-4579-b31a-b5c6e9632b04"
      * system = "urn:hl7-ru:fias:aoguid"
  * extension[houseguid]
    * valueIdentifier
      * value = "6a5dfa69-aeb3-4422-82a8-3a577f5e36a6"
      * system = "urn:hl7-ru:fias:houseguid"
```

**Reply by eugenekogan-spb** [reply]
Результат: расхождений в нотации нет, с точки зрения состава информации целесообразно добавить в Core Address extension. Предположительно 

regionRF 1..1 coding, заполняется по справочнику субъектов РФ, 1.2.643.5.1.13.13.99.2.206

**Comment by eugenekogan-spb** [comment]


**Reply by eugenekogan-spb** [reply]
2.1.5. Указание на отсутствие информации NullFlavor

<img width="658" alt="image" src="https://github.com/user-attachments/assets/a188005c-d89d-48c3-8a4b-bf6ee9fd4694">

**Reply by eugenekogan-spb** [reply]
Хотелось бы привести здесь предлагаемый FHIR подход к NullFlavor - он описан в этом руководстве, п.1.1.: http://build.fhir.org/ig/HL7/ccda-on-fhir/

> The Composition profiles in this IG do not require coded data in any section. This is a departure from C-CDA, which requires coded data for Problems, Results, Medications, etc. This departure is intentional, as the C-CDA requirement for one or more coded entries in these sections resulted in some very complicated workarounds using nullFlavors to handle the fact that sometimes a patient is not on any medications, or has no current problems. In general, FHIR takes the approach that if something is nullable, it should simply be optional to ease the burden on implementers, thus C-CDA on FHIR does not require any coded entries, but rather uses the “required if known” approach, meaning that if an implementer’s system has data for a section that requires data under Meaningful Use, they need to send it, but if they have no data there is no need for a null entry.

**Reply by Vrodvi** [reply]
На основании разделов руководства по реализации СЭМД:
```
2.1.5 Указание на отсутствие информации <nullFlavor>
Если информация для элементов, которые должны присутствовать, неизвестна, то
факт отсутствия этой информации должен быть отражен с использованием
подходящего значения атрибута nullFlavor.

### 5.1.2 Ключевые слова

Ключевые слова, использованные в данном руководстве: "ОБЯЗАН"/"НЕ
ОБЯЗАН", "ДОЛЖЕН"/"НЕ ДОЛЖЕН", "СЛЕДУЕТ"/"НЕ СЛЕДУЕТ", "МОЖЕТ"/"НЕ
МОЖЕТ".

"ОБЯЗАН"/"НЕ ОБЯЗАН" – обязательное требование включения/исключения
элемента. Нарушение приведет к несоответствию сформированного CDA документа
требованиям данного руководства. Ключевое слово "ОБЯЗАН"/"НЕ ОБЯЗАН"
используется для описания варианта, когда элемент НЕ может иметь пустое наполнение
и НЕ может содержать атрибут nullFlavor.

"ДОЛЖЕН"/"НЕ ДОЛЖЕН" – обязательное требование включения/исключения
элемента. Нарушение приведет к несоответствию сформированного CDA документа
требованиям данного руководства. Ключевое слово "ДОЛЖЕН"/"НЕ ДОЛЖЕН"
используется для описания варианта, когда элемент не может иметь пустое наполнение,
но может содержать атрибут nullFlavor.

Атрибут nullFlavor используется для обозначения причины отсутствия
информации и может быть использован в любом элементе документа, который задается
ключевым словом "ДОЛЖЕН". Значение атрибута nullFlavor определяется согласно
справочнику "Причины отсутствия информации (NullFlavor)" (OID:
1.2.643.5.1.13.13.99.2.286). В требованиях текущего руководства по реализации
используется следующая логика по умолчанию:
• Если элемент не содержит атрибут nullFlavor, то все указанные в требованиях
атрибуты обязательны к указанию в xml-документе;
• Если элемент содержит атрибут nullFlavor, то все атрибуты, представленные не
пустыми значениями, должны быть представлены, все атрибуты с пустыми
значениями могут отсутствовать.

Если потребуется использовать логику, отличную от указанной выше, она будет
описана в требованиях дополнительно.

В XML примерах можно встретить следующие обозначения данных требований:

R [1..1] – данный элемент должен быть представлен с не пустым значением, т.е.
использование атрибута nullFlavor запрещено. При отсутствии значения в таких
обязательных элементах документ CDA не может быть передан в РЭМД.

[1..1] – данный элемент должен быть представлен, но значение может
отсутствовать при указании причины отсутствия информации в атрибуте nullFlavor.
Данный элемент может присутствовать только в одном экземпляре.

[1..*] – данный элемент должен быть представлен, но значение может
отсутствовать при указании причины отсутствия информации в атрибуте nullFlavor.
Данный элемент может присутствовать в неограниченном количестве, но минимум в
одном экземпляре.

```
Атрибут `nullFlavor` может указываться в любых элементах, его использования явно в элементах не оговаривается. Имеется косвенное требование на использованием атрибута `nullFlavor` через ключевое слово `ДОЛЖЕН` в описании требований к элементу.

При этом атрибут `nullFlavor` не обязательно полностью заменяет собой всё наполнение элемента (об этом свидетельствует фрагмент `Если элемент содержит атрибут nullFlavor, то все атрибуты, представленные не
пустыми значениями, должны быть представлены, все атрибуты с пустыми
значениями могут отсутствовать`), как например вот тут где нужно указать в элементе код с типизацией, но основное наполнение элемента уже подавляется через указанный `nullFlavor`: 
```
<!-- [1..1] Значение поля -->
<value xsi:type="CD" nullFlavor="NI"/>
```

---

Если ориентироваться на подход из руководства по реализации СЭМД, то можно аналогичным образом разрешить повсеместное использование стандартного расширения https://hl7.org/fhir/extensions/StructureDefinition-iso21090-nullFlavor.html

Пример наполнения "организации", взятой из СЭМД, переложенной в FHIR (FSH) с использованием расширения nullFlavor на всех "ДОЛЖНЫХ" элементах:
```
Instance: organization-instance
InstanceOf: Core_Organization
Title: "Organization Instance"
Description: "ОБЩЕСТВО С ОГРАНИЧЕННОЙ ОТВЕТСТВЕННОСТЬЮ \"ХЕЛИКС АЛТАЙ\""
Usage: #example

//			<!-- R [1..1] Наименование организации или ФИО ИП -->
//			<name>ОБЩЕСТВО С ОГРАНИЧЕННОЙ ОТВЕТСТВЕННОСТЬЮ "ХЕЛИКС АЛТАЙ"</name>

* name = "ОБЩЕСТВО С ОГРАНИЧЕННОЙ ОТВЕТСТВЕННОСТЬЮ \"ХЕЛИКС АЛТАЙ\""

//			<!-- R [1..1] Адрес медицинской организации или ИП -->
//			<addr>

* address
  * use = #work
  * type = #postal

//				<!-- R [1..1] Адрес (текстом) -->
//				<streetAddressLine>Алтайский край, город Барнаул, улица Сергея Ускова, дом 33</streetAddressLine>

  * text = "Алтайский край, город Барнаул, улица Сергея Ускова, дом 33"

//				<!-- [1..1] Почтовый индекс -->
//				<postalCode nullFlavor="NI"/>

  * postalCode
    * extension
      * url = "http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor"
      * valueCode = #NI

//				<!-- R [1..1] Код субъекта РФ -->
//				<address:stateCode codeSystem="1.2.643.5.1.13.13.99.2.206" codeSystemName="Субъекты Российской Федерации" code="22" codeSystemVersion="6.4" displayName="Алтайский край"/>
  
  * extension[0]
    * url = "urn:hl7-ru:regionRF"
    * valueCoding
      * code = #22
      * system = "1.2.643.5.1.13.13.99.2.206"
      * version = "6.4"         
      * display = "Алтайский край"
  
//				<!-- [1..1] Идентификатор адреса по ФИАС -->
//				<fias:Address nullFlavor="NI"/>  

  * extension[fias]
  
    * extension[0]
      * url = "http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor"
      * valueCode = #NI

//			<!-- R [1..1] Уникальный идентификатор медицинской организации и (при наличии) уникальный идентификатор структурного подразделения -->
//			<id root="1.2.643.5.1.13.13.12.2.22.27143"/>
 
* identifier[FRMO]
  * system = "http://fhir.ru/core/systems/frmo"
  * value = "1.2.643.5.1.13.13.12.2.22.27143"
  
//			<!-- [0..1] Телефон медицинской организации или ИП -->
//			
  
* contact
  * telecom[0]
    * system = #phone
    * value = "(+7) 909-502-7809"
```

Такой пример корректно обрабатывается в FSH Online и Publisher тоже его обрабатывает, хотя в этом примере есть принципиальная ошибка с обозначением кардинальности вложенных элементов в элементы, которым разрешается указание nullFlavor - если будет адекватный валидатор, который строго проверит установленные в профилях крадинальности, то фрагмент с ФИАС проверку не должен пройти, т.к. для расширения ФИАС обязательно должен быть указан вложенный элемент  aoguid. FSH Online и Publisher такую валидацию не проводят, но выдают предупреждение о том, что без явного указания срезов в расширении проверка кардинальности может выполниться неправильно (какой смекалестый!):
```
warn  Element Organization.address.extension:fias.extension:aoguid.value[x] has its cardinality satisfied by a rule that does not include the slice name. Use slice names in rule paths when possible.
``` 

Чтобы такого не было, придётся во все места, где возможно отсутствие сведений (и следовательно, возможно указание nullFlavor), убирать кардинальность до необязательной `0..`,  а также добавлять сложные правила кастомных проверок (`Invariants`), которые должны будут обозначить необходимость указания элементов через ИЛИ - или указание nullFlavor или укзанием основного наполнения.
Если посмотреть внимательнее на пример из СЭМД, то в том месте, где якобы идёт обязательность `1..`, но разрешён nullFlavor, на самом деле кардинальность явно прописана как '0..', и даётся правило на выбор либо nullFlavor либо наполнения:

вот как выглядит в руководстве (якобы кардинальность `1..`):
![image](https://github.com/user-attachments/assets/4eb8abff-d634-44e3-8359-ac5e6be7e0cc)

вот как выглядит в фактической схеме проверки (всё опциональное, но есть правило или nullFlavor или наполнение):
![image](https://github.com/user-attachments/assets/4266efa3-774a-4d39-a752-3bd9cc139873)


----

Теперь возвращаемся к альтернативному варианту работы с укзаанием отсутствия данных из руководства http://build.fhir.org/ig/HL7/ccda-on-fhir/
В нём, как раз таки явно написано, что использование nullFlavor с обязательн включением элементов без наполнения приводит к необходимости построения очень сложных обходных конструкций (`resulted in some very complicated workarounds using nullFlavors`) - что и видно на примере с ФИАС.
Вместо использования nullFlavors в руководстве ccda-on-fhir решено сделать все элементы, в которых может отсутствовать информация опциональными и не усложнять жизнь тем, кто будет работать над релаизацией руководства. 
Подход такой - если есть информация, то укажи, а если информации нет, то не нужно указывать факт её отсутствия.

----

Таким образом, явное использование nullFlavor как в СЭМД возможно, но потребует больших усилий по написанию комплексных правил проверки в профилях ресурсов. 

Альтернативный подход - не использовать nullFlavor, сделав все атрибуты, где может отстутсовать информация, опциональными. (В этом случае при необходимости конвертации FHIR в CDA придётся предусматривать механизм восстановления пропущенных элементов и установки в них атрибута nullFlavor как часть процесса конвертации данных)

---- 

Дальнейший анализ по этой теме требует обсуждения и принятия одного из подходов.

**Reply by eugenekogan-spb** [reply]
Предлагаю обсудить раздельно две ситуации, технически одинаково выраженные через NullFlavor, но, на мой взгляд, принципиально различные. 1) значения точно нет, 2) у нас нет информации о значении.

Вариант 1. Атрибуты, у которых значение может быть, а может отсутствовать.
Примеры: у пациента нет отчества. Пациент не принимает медикаменты. При передаче через NullFlavor видимо это NA
<img width="932" alt="image" src="https://github.com/user-attachments/assets/c2623cf3-c129-4db3-bd93-174680212602">

На мой взгляд, это значимая информация. Ее можно передавать через NullFlavor, но можно моделировать отдельным расширением, как это часто делают при заполнении личных данных: делают чек-бокс "Отчества нет".

Вариант 2. Атрибут, у которого значение есть всегда, но мы по какой-то причине не указали.
Пример: почтовый адрес организации. Если поле осталось пустым, это не может означать, что адреса нет.  

Для этого случая NullFlavor предлагает много вариантов: у пациента не спросили, у пациента спросили но он не ответил, значения сейчас нет но оно будет позже и т.д. На мой взгляд, эта модель не получила реального использования. Мне не попадался ни один интерфейс пользователя, позволяющий выбрать такие значения. Вместо этого разработчики хардкодят nullFlavor = NI (No Information).

Итого, предлагаю
1) Для тех немногих атрибутов, где значение может реально отсутствовать, и его отсутствие представляет собой осмысленную информацию, предлагаю эту информацию моделировать явно, а при передаче FHIR->CDA превращать в NullFlavor = NA 
2) Для остальных атрибутов, где спецификация CDA устанавливает "должен", ставить кардинальность 0.. , а при отсутствии значения передавать FHIR->CDA в виде NullFlavor = NI

**Reply by Vrodvi** [reply]
Для "1) значения точно нет", скорее всего нужно использовать вообще другое расширение - https://hl7.org/fhir/extensions/StructureDefinition-data-absent-reason.html

NullFlavor больше похоже на обработку ситуации "2) у нас нет информации о значении."

**Reply by eugenekogan-spb** [reply]
Data-absent-reason выглядит как калька NullFlavor
<img width="1102" alt="image" src="https://github.com/user-attachments/assets/a01d8937-1ad8-4126-bc06-a3b277bc8402">

**Reply by Vrodvi** [reply]
Нет, насколько я понял они принципиально по подходу отличаются.

*NullFlavor*
- вставляется вместо значения, т.е. это тоже вариант значения некого атрибута:
```
* name
  * family = "Габсбург-Лотарингская"
  * given[0] = "Мария-Антуанетта"
  * given[1] = тут либо Отчество либо расширение NullFlavor 
```


*Data-absent-reason*
- вставляется в дополнение к осноному "пустому" значения, для того чтобы уточнить почему само значение "пустое":
```
* name
  * family = "Габсбург-Лотарингская"
  * given[0] = "Мария-Антуанетта"
  * given[1] = тут либо пустота либо что-то вроде "Отчество отсутствует"
    * тут расширение Data-absent-reason
```

**Reply by Vrodvi** [reply]
Этот пункт остался без чёткого решения.

Составить пример с использованием стандартного расширения nullFlavor не сложно, сложно будет его кому-то потом реализовывать, описывать профили под вариации в наполнении элементов и осуществлять валидацию экземпляров на соответствие усложнённых конструкций, описанных в профиле.

Некоторые моменты которые могут помочь разрешить ситуацию:
- на еженедельном совещании было озвучено, что элементы с требованием "ДОЛЖЕН" встречаются не часто, и, если это так, то небольшое количество мест с требованием "ДОЛЖЕН" можно прицельно рассмотреть и подобрать там наиболее подходящий формализованный варинат описания отсутсвия данных. Требуется сбор статистики по количеству требований "ДОЛЖЕН".
- на еженедельном совещании было озвучено, что стремления к точному соответствию 1-к-1 между структурами СЭМД и FHIR нет, поэтому структуры могут отличаться существенно, а обязанности по правильной конвертации между структурами СЭМД и FHIR можно будет перелжить на некий отдельный процесс/конвертер, который на данном этапе рассматриваться не будет.

В данных условиях буду считать вариантом по-умолчанию следующий:
`Для атрибутов, где спецификация CDA устанавливает "должен", ставить кардинальность 0..`

**Reply by eugenekogan-spb** [reply]
Если принять в точности это решение, то позиции спецификации CDA "должен" и "может" будут в FHIR кодироваться одинаково. В этом случае "процесс/конвертер" не сможет правильно сконвертировать ввиду недостатка информации.

Предлагаю 
Для атрибутов, где спецификация CDA устанавливает "должен", ставить кардинальность 0.. и указывать возможность NullFlavor.

**Comment by eugenekogan-spb** [comment]
2.1.4. Дата и время

<img width="640" alt="image" src="https://github.com/user-attachments/assets/3889622d-c2fc-439b-be4c-4c400589e6a3">

**Reply by Vrodvi** [reply]
В CDA (СЭМД) для указания даты и времени используется международный стандарт ISO 8601.
Даты указываются в формате YYYYMMDDhhmmss.SSS±ZZzz.

В FHIR имеется два типа данных:
date - указание только даты в формате YYYY-MM-DD
dateTime - дата со временем в формате YYYY-MM-DDThh:mm:ss+zz:zz

Таким образом CDA и FHIR будут совпадать в плане указания даты/времени при использовании типа данных dateTime (только немного отличается формат записи - в FHIR используются разделители, а в CDA - нет), а при использовании в FHIR типа date - будут частично совпадать.

Проблемные места будут там, где FHIR использует тип date, а CDA полный формат. Как минимум такое место есть при указании даты рождения пациента. В СЭМД для детей до года дата рождения указываетя с точностью до минут, а FHIR дата рождения имеет тип date. В этом месте в FHIR придётся использовать стандартное расширение для указания даты/времени рождения - 
https://hl7.org/fhir/extensions/StructureDefinition-patient-birthTime.html для полного соответствия СЭМДу.

Исходный фрагмент из СЭМД:
```
<!-- [1..1] Дата и время рождения пациента -->
<birthTime value="19791125"/>

<!-- R [1..1] Дата измерения -->
<effectiveTime value="202211150830+0300"/>
```
Переложение в FHIR:
```
Instance:   core-patient-instance-full
InstanceOf: Patient

* birthDate = "1979-11-25"
  * extension[0]
    * url = "http://hl7.org/fhir/StructureDefinition/patient-birthTime"
    * valueDateTime = "1979-11-25T15:30:10+01:00"


Instance:   core-observation-instance-full
InstanceOf: Core_Observation

* effectiveDateTime = "2022-11-15T08:30:00+03:00"
```

---
---

Для анализа использованы следующие фрагменты руководства СЭМД:

---

2.1.4 Дата и время

Времена или временные интервалы событий, описываемые/упоминаемые в
документе, должны быть указаны с точностью до дня (если не указано иное), но
желательно указывать с точностью до минут. Если время указано с точностью до минут,
то должна быть указана временная зона (часовой пояс). Пример указания даты и
времени представлен ниже (см. Рисунок 10).

1. Дата и время должны быть представлены в формате
"YYYYMMDD[HHMM[SS]+|ZZZZ]", где "YYYY" – год, "ММ" – месяц, "DD" – день,
"HHMM[SS]" – время ("HH"- часы, "MM" - минуты, "SS" - секунды), "+|ZZZZ" –
указатель временной зоны (квадратные скобки указывают на элементы, которые
можно опустить). Временные интервалы должны быть указаны с точностью как
минимум до дня, если иное не предусмотрено разработчиком (для некоторых
документов допустимо указание временного интервала с точностью до года).
Необходимость указания времени с большей точностью, чем до дня, оговаривается в
требованиях отдельно. Допускается указание даты и времени с максимальной
возможной точностью (до секунд).
```
<!-- 5 марта 2021 года, 17:15:40, часовой пояс Москвы (UTC+03:00) -->
<effectiveTime value="20210305171540+0300"/>
```
Рисунок 10. Пример указания даты и времени.

---

2.8 Дата создания документа <effectiveTime>

Дата создания документа указывается в элементе effectiveTime с точностью как
минимум до дня, рекомендуется указывать с точностью до минут (секунд) с указанием
временной зоны.
Пример описания компонента представлен на Рисунок 19.
У1-7: Элемент ClinicalDocument ДОЛЖЕН иметь один [1..1] элемент effectiveTime
(атрибут @value должен быть представлен с не пустым значением), содержащий дату
создания документа. ЕСЛИ документ подписывается посредством УКЭП, ТО атрибут
effectiveTime/@nullFlavor должен иметь значение "NA".
```
<!-- [1..1] Дата создания документа -->
<effectiveTime nullFlavor="NA"/>
```
Рисунок 19. Пример указания effectiveTime.

---

Примеры:
```

6. ОБЯЗАН иметь один [1..1] элемент identity:IssueDate (атрибут @value
должен быть представлен с не пустым значением), содержащий дату выдачи
документа;

<!-- R [1..1] Дата выдачи документа -->
<identity:IssueDate value="20050218"/>

---

3. ДОЛЖЕН иметь один [1..1] элемент birthTime (атрибут @value должен
быть представлен с не пустым значением), содержащий дату и время
рождения пациента (для взрослых и детей старше одного года должен быть
указан с точностью до дня, для детей до года следует указывать с точностью
до минут (допускается до секунд));

<!-- [1..1] Дата и время рождения пациента -->
<birthTime value="19791125"/>

---

2. ОБЯЗАН иметь один [1..1] элемент effectiveTime (атрибут @value должен
быть представлен с не пустым значением), содержащий дату формирования
протокола лабораторного исследования;

<!-- R [1..1] Дата формирования протокола лабораторного исследования -->
<effectiveTime value="202210312150+0400"/>

---

<!-- [1..1] Срок действия документа-основания -->
<identity:effectiveTime>
<!-- R [1..1] Дата начала действия документа-основания -->
<identity:low value="20171025"/>
<!-- [1..1] Дата окончания действия документа-основания -->
<identity:high nullFlavor="NAV"/>
</identity:effectiveTime>

```

**Comment by eugenekogan-spb** [comment]
2.1.6. Уникальные идентификаторы <Id>

<img width="670" alt="image" src="https://github.com/user-attachments/assets/443772bf-5052-451c-ab30-1a31f2facf4e">

**Reply by Vrodvi** [reply]
В Ru Core для пространст идентификаторов используются URL, в то время как для СЭМД обязательно использования OID.

Более простая ситуация тут с логическими идентификаторами (СНИЛС, свидетельство о рождении, и т.п.).
Для этого блока идентификаторов для обеспечения возможности конвертации из FHIR в СЭМД необходимо, чтобы все пространства идентификатор, используемые в FHIR были дополнены вспомогательными значениями OID.
При этом значения OID нельзя сгенерировать случайным образом - их необходимо получить через владельца одного из существующих узлов OID (например, через узел Минздрава РФ - https://nsi.rosminzdrav.ru/oid)

На примере пациента, OIDы отсутствуют у следующих пространств идентификаторов:
- https://build.fhir.org/ig/fhir-ru/RuCoreIG/NamingSystem-core-ns-rf-inn.html
- https://build.fhir.org/ig/fhir-ru/RuCoreIG/NamingSystem-core-ns-rf-birth-certificate.html
- https://build.fhir.org/ig/fhir-ru/RuCoreIG/NamingSystem-core-ns-rf-snils.html

Более сложная ситуация с внутренними идентификаторами FHIR-ресурсов. OID для них заранее завести не получится. Эти OID нужно формировать по правилам описываемым в разделе "5.4 Использование идентификаторов", например для идентификации пациентов "Идентификаторы пациентов в экземпляре МИС в медицинской рганизации" - нужно формировать OID по фасету "OID_медицинской_организации.100.Номер МИС.НомерЭкзМИС.10"

В случае наличия OID значений в описании пространств идентификаторов сами экземпляры FHIR из Core руководства можно оставить без изменения.

Пример:

 - сам пациент остаётся без изменения:
```
Instance:   core-patient-instance-full
InstanceOf: Core_Patient
Title: "Core Instance Patient Full"
Description: "Пациент со свидетельством о рождении"
Usage: #example

* identifier[birthCertificate]
  * system = "http://fhir.ru/core/systems/birth-certificate"
  * value = "IIЮБ-123456"
  * type = Core_Cs_Nsi_Identity_Document#6
```

- в пространстве идентификаторов свидетельств о рождении нужно добавить OID:
```
Instance:   core-ns-rf-birth-certificate
InstanceOf: NamingSystem
Title: "Core NamingSystem RF Birth Certificate (Номер гос свидетельства о рождении)"
Usage: #definition

* name =   "Core_Ns_Rf_Birth_Certificate"

* description = "Номера государственных свидетельств о рождении формируются государственной системой органов записи актов гражданского состояния (ЗАГС), с внесением в единый реестр записей актов гражданского состояния (ЕГР ЗАГС). Предполагается, что номер уникален, сведений об ограничении срока уникальности номера нет."

* status = #active

* kind = #identifier

* date = "2023-11-22"

* uniqueId[0]
  * type = #uri
  * value = "http://fhir.ru/core/systems/birth-certificate"
  * preferred = true
  
* uniqueId[1]
  * type = #oid
  * value = "нужно получить OID узел"
```

Пространство идентификаторов для внутреннего идентификатора (в примере - "core-patient-instance-full") в автоматическом режиме заменить не получится - это нужно зафиксировать как известную проблему конвертации между СЭМД и FHIR.
Даже если МИС заведёт у себя записи о пространствах идентификаторов для всех внутренних идентификаторов ресурсов, то всё равно нужно будет иметь дополнительную жёстко зашитую логику по их использованию, т.к. в экземпляре FHIR ресурса пространство идентификаторов для внутренних идентификаторов вообще не указываестся.

**Reply by Vrodvi** [reply]
Проблема с внутренними идентификаторами известная, в CCDA-on-FHIR с ней также столкнулись:
https://build.fhir.org/ig/HL7/ccda-on-fhir/mappingGuidance.html#fhir-ids
![image](https://github.com/user-attachments/assets/33ddfe6f-715e-42ca-abe9-e05d912fed63)

**Reply by Vrodvi** [reply]
Обсудили, что все динамически формируемые пространства идентификаторов, указанные в разделе "5.4 Использование идентификаторов" нужно заводить в FHIR как дополнительные "бизнес-идентификаторы" (в атрибуте identifier).

В таком случае, к примеру, "Идентификаторы пациентов в экземпляре МИС в медицинской организации" (формируются по фасету "OID_медицинской_организации.100.Номер МИС.НомерЭкзМИС.10") можно представить в FHIR следующим образом:

исходно в СЭМД:
```
			<!-- R [1..1] Уникальный идентификатор пациента в МИС -->
			<id root="1.2.643.5.1.13.13.12.2.22.1576.100.1.1.10" extension="3421"/>
```
представление в FHIR:
```
Instance:   core-patient-instance-full
InstanceOf: Core_Patient
Title: "Core Instance Patient Full"
Description: "Пациент со свидетельством о рождении"
Usage: #example

* identifier[Идентификатор пациента в экземпляре МИС в медицинской организации]
  * system = "1.2.643.5.1.13.13.12.2.22.1576.100.1.1.10"
  * value = "3421"
  * type = http://terminology.hl7.org/CodeSystem/v2-0203#PRN
```
, где:
-  код #PRN - это наиболее подходящий код типа идентификатора из рекомендованного справочника - код "поставщика медицинских услуг" (оригинальное описание - https://terminology.hl7.org/5.1.0/CodeSystem-v2-0203.html#v2-0203-PRN `Provider number | A number that is unique to an individual provider, a provider group or an organization within an Assigning Authority.`)
- пространство `1.2.643.5.1.13.13.12.2.22.1576.100.1.1.10` - сформировано динамически по фасету `OID_медицинской_организации.100.Номер МИС.НомерЭкзМИС.10`

**Reply by eugenekogan-spb** [reply]
@Vrodvi Виталий, правильно ли я понимаю, что при таком подходе тип #PRN следует применять для всех идентификаторов, выданных медицинской организацией, в частности, номеров медицинских карт, случаев обслуживания и т.п.?

**Reply by Vrodvi** [reply]
Нет.
Что-то я тут напутал.
#PRN - это подходит для идентификатора саой МО, а для идентификатора пациента внутри неё, нужен другой тип.
Более подходящим, наверное будет "PI"	Patient internal identifier.
Для других типов идентификаторов нужны другие коды, для медицинской карты - "MR"	Medical record number, для случаев обслуживания - сходу подходящего кода не нашлось.
Но тут ещё интересный момент, что тип идентификатора часто вообще не нужен, т.к. "пространство идентификаторов" его однозначно определяет.

**Comment by eugenekogan-spb** [comment]
2.1.3. Контакты
<img width="624" alt="image" src="https://github.com/user-attachments/assets/0574ac64-8008-4354-9d75-cecb1e4dbe92">

<img width="635" alt="image" src="https://github.com/user-attachments/assets/7de6171e-d0f5-4c02-bd64-c17c75dd3637">

<img width="642" alt="image" src="https://github.com/user-attachments/assets/5465fee7-851f-404c-bf84-e8b24d2fbcb6">

<img width="646" alt="image" src="https://github.com/user-attachments/assets/0655104f-0e3a-47cc-bad5-bc681d806c80">

**Reply by eugenekogan-spb** [reply]
Анализ
Тип данных ContactPoint позволяет указать в элементе system варианты #phone, #fax, #email
В строке value можно заполнять значение с префиксом, как это рекомендует спецификация

Пример может выглядеть следующим образом:
```
* contactPoint
   * system = #phone
   * value = "+7(912)345-67-89"
* contactPoint
   * system = #email
   * value = "name@mail.ru"
* contactPoint
   * system = #fax
   * value = "7-095-345-6789"
```

**Reply by eugenekogan-spb** [reply]
Вывод: информация взаимно однозначна, но нотация различается: при переводе в CDA необходимо добавлять в value префикс, зависимый от system.

**Reply by eugenekogan-spb** [reply]
Дополнение (О.Пензин, встреча 17.01.25):
везде добавляем 
```
* use = #work
* rank = (номер по порядку, начиная с 1)
```
